File: Online-judge/leetcode/maximum-level-sum-of-a-binary-tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/maximum-level-sum-of-a-binary-tree/description/
 * Time complexity: O(n).
 * Space complexity: O(n) stack for recursion call
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
private:
    void solve(TreeNode* root, int level, vector<int>& res) {
        if (!root) return;
        if (res.size() == level) {
            res.push_back(root->val);
        } else {
            res[level] += root->val;
        }
        
        solve(root->left, level + 1, res);
        solve(root->right, level + 1, res);
    }

public:
    int maxLevelSum(TreeNode* root) {
        vector<int> res;
        this->solve(root, 0, res);
        int mx = INT_MIN;
        int index = 0;

        for (int i = 0; i < res.size(); i++) {
            if (res[i] > mx) {
                mx = res[i];
                index = i;
            }
        }
        return index + 1;
    }
};

File: Online-judge/leetcode/maximum-level-sum-of-a-binary-tree.py
# Definition for a binary tree node.
import collections
from typing import Optional

"""
Md.Shohag <shohag.fks@gmail.com>
https://leetcode.com/problems/maximum-level-sum-of-a-binary-tree/description/
Time complexity O(N)
Space complexity O(n) Tree Hight 2^h = N
"""

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        mx, ans, level = float("-inf"), 0, 0
        q = collections.deque()
        q.append(root)

        while q:
            level += 1
            currentLevelSum = 0
            qsize = len(q)

            for _ in range(qsize):
                top = q.popleft()
                currentLevelSum += top.val
                if top.left:
                    q.append(top.left)
                if top.right:
                    q.append(top.right)
                    
            if currentLevelSum > mx:
                mx = currentLevelSum
                ans = level
        return ans